performance-es.md

```diff
@@ -183,3 +185,107 @@ de texto, pero el coste es que no será legible para un humano.
+
+### Cambiar de algoritmo
+
+Si no estás cambiando los datos, la alternativa más importante es cambiar el código.
+
+Es muy posible que las mejoras más importantes vengan de un cambio de algoritmo. Esto es equivalente a sustituir bubble sort (`O(n^2)`) con quicksort (`O(n log n)`) o reemplazar un acceso linear a un array (`O(n)`) con una busqueda binaria (`O(log n)`) o una busqueda en un mapa (`O(1)`).
+
+Así es como el software se vuelve lento. Estructuras originalmente diseñadas para un proposito se reusan para algo que no habian sido diseñadas. Esto ocurre gradualmente.
+
+Es importante tener un entendimiento intuitivo de los diferentes niveles de big-O. Elige la estructura de datos para tu problema. Esto no siempre ahorra ciclos de CPU, pero previene problemas de rendimiento que pueden no ser detectados hasta muchos más adelante.
+
+Las clases básicas de complejidad son:
+
+* O(1): acceso a un campo, un array o un mapa
+
+  Consejo: no te preocupes por ellos
+
+* O(log n): busqueda binaria
+
+  Consejo: sólo es un problema si se hace en un bucle
+
+* O(n): bucle simple
+
+  Consejo: lo haces todo el tiempo
+
+* O(n log n): divide y vencerás, ordenación
+
+  Consejo: sigue siendo bastante rapido
+
+* O(n\*m): bucles anidados / cuadrático
+
+  Consejo: ten cuidado y limita el tamaño de tu conjunto de datos
+
+* Cualquier cosa entre cuadrático y subexponencial
+
+  Consejo: no lo ejecutes en un millón de filas
+
+* O(b ^ n), O(n!): exponencial y mayor
+
+  Consejo: vas a necesitar suerte si tienes más de una o dos docenas de datos
+
+Link: <http://bigocheatsheet.com>
+
+Supongamos que tienes que buscar en un conjunto desordenado de datos. "Debería usar busqueda binaria" piensas, sabiendo que una busqueda binaria es O(log n) que es más rapido que el O(n) de una busqueda linear. Sin embargo, una busqueda binaria requiere que los datos estén ordenados, lo que significa que tendrás que ordenarlos antes, que tarda O(n log n). Si haces muchas busquedas, el coste inicial de la ordenación merecerá la pena. Pero, si sobre todo estas haciendo **lookups**, quizás usar un array fue una decisión equivocada y sería mejor usar un mapa con coste O(1).
+
+Si tu estructura de datos es estática, entonces generalmente podrás hacerlo mucho mejor que en el caso de que fuera dinámica. Resultará más facil construir una estructura de datos óptima para tus patrones de busqueda. Soluciones como minimal perfect hashing pueden tener más sentido aquí, o filtros de Bloom precalculados. Esto también tiene sentido si tu estructura de datos es "estática" durante un periodo largo de manera que puedas amortizar el coste inicial de su construcción en muchas busquedas.
+
+Escoje la estructura de datos más simple que sea razonable y continua. Esto es de primero de carrera para escribir "software no lento". Este debe ser tu modo de desarrollar por defecto. Si sabes que necesitas acceso aleatorio, no escojas una lista enlazada. Si sabes
```

Comment:
`/s/continua/continúa`

Trying to come up with an accurate `CS 101` translation in Spanish. We might have to change a bit of the wording to make this translation make sense in Spanish, given that the original meaning, the fact that this is a basic/beginner level recommendation, is still portrait in this phrase.

 Something about `Esto es de primero de carrera para escribir "software no lento"` does not sound so good to me.

Reply:
Agreed, let's figure out a better translation.

Reply:
I have applied @martskins suggestion on Slack and translated `This is CS 101 for
writing "not-slow software"`as `Esto es elemental para escribir "software no lento"`.